mobile_payment/utils.py

```diff
@@ -0,0 +1,76 @@
+import datetime
+import requests
+import logging
+from datetime import datetime, timedelta
+from django.conf import settings
+from django.utils.translation import gettext_lazy as _
+import json
+from .models import ApiUtilitie
+import threading
+
+logger = logging.getLogger(__name__)
+
+lock = threading.Lock()
+
+def access_token():
+    # make it thread safe
+    with lock:
+        try:
+            NOW = datetime.now()
+            api_name = 'Qmoney_token'
+            access_token = get_access_token(api_name)
+            return {"Authorization": f"Bearer {access_token}"}
+        except Exception as e:
+            logging.exception("failed to get access token")
+            return e
+
+
+def get_access_token(api_name):
+    api_utility = ApiUtilitie.objects.get(name=api_name)
+    current_token = api_utility.access_token
+    token_expiry_date = api_utility.access_TokenExpiry
+
+    if token_expiry_date is None or token_expiry_date <= datetime.now():
+        new_access_token, new_expiry_date = requests_new_access_token(api_name)
+
+        api_utility.access_token = new_access_token
+        api_utility.access_TokenExpiry = new_expiry_date
+        api_utility.save()
+
+        return new_access_token
+
+    return current_token
+
+
+def requests_new_access_token(api_name):
+    try:
+        auth_url =settings.PSP_QMONEY_AUTH_URL
+        grantype=settings.PSP_QMONEY_GRANTTYPE
+        username = settings.PSP_QMONEY_USERNAME
+        password = settings.PSP_QMONEY_PASSWORD
+        authBearerToken=settings.PSP_QMONEY_AUTH_BEARER_TOKEN
```

Comment:
could you please explain me why you need this token? Isn't the username and password enough?

Reply:
This is one of the Requirement by Qmoney for us to be able to generate an Access Token

Reply:
I got access to a part of their API doc _QCell MobiFin Elite 5.0 API Guide - Merchant_.

In that doc, I confirm that this authorization header is mentioned and mandatory. This is unusual, hence my question.

so it's done.